### PR TITLE
match lowering: eagerly simplify match pairs

### PR DESCRIPTION
**Pull request by Nadrieril**
This removes one important complication from match lowering. Before this, match pair simplification (which includes collecting bindings and type ascriptions) was intertwined with the whole match lowering algorithm.

I'm avoiding this by storing in each `MatchPair` the sub-`MatchPair`s that correspond to its subfields. This makes it possible to simplify everything (except or-patterns) in `Candidate::new()`.

This should open up further simplifications. It will also give us proper control over the order of bindings.

r? @matthewjasper

### CODE REVIEW

File: compiler/rustc_mir_build/src/build/matches/test.rs [hunk 1]
```diff
@@ -589,22 +589,17 @@ impl<'a, 'tcx> Builder<'a, 'tcx> {
         // away.)
         let (match_pair_index, match_pair) =
             candidate.match_pairs.iter().enumerate().find(|&(_, mp)| mp.place == *test_place)?;
+        let mut fully_matched = false;
```
Comment (matthewjasper) [inline compiler/rustc_mir_build/src/build/matches/test.rs:592]: I would prefer this to be `let fully_matched;` and have every branch assign true or false.

**Review by matthewjasper — COMMENTED** [review]
Couple of style/comment nits.

File: compiler/rustc_mir_build/src/build/matches/test.rs (hunk 1 above)
Comment (Nadrieril) [reply compiler/rustc_mir_build/src/build/matches/test.rs:592]: Ooh yeah, in another branch I in fact forgot to set it once, and the resulting infinite loop crashed my laptop x) Good shout

**Review by matthewjasper — COMMENTED** [review]
@bors r+

**Comment by Nadrieril** [comment]
Possibly spurious

@bors retry

**Comment by Nadrieril** [comment]
I can't tell which is the test that timed out or if this has anything to do with this PR.

@bors retry

**Comment by rylev** [comment]
Improvements far outweigh the regressions so I don't think this needs further investigation.

@rustbot label: +perf-regression-triaged

**Pull request merged**